### PR TITLE
Don't constrain projection predicates with inference vars in GAT substs

### PR DESCRIPTION
**Pull request by jackh726**
cc #91762 

Not a fix, but a mitigation to prevent a backwards-compatible hazard where we normalize using a predicate only because it's the only one available, but shouldn't. This would constrain an inference variable which didn't really want. We already do this when selecting a projection candidate, which isn't always correct. But changing that is a problem for a different day.

Also found out that a suggestion for `await`ing a future was using the wrong substs.

r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 1]
```diff
@@ -1082,6 +1082,20 @@ fn project_type<'cx, 'tcx>(
...
     }
 
+    // If the obligation contains any inference types or consts in associated
+    // type substs, then we don't assemble any candidates.
+    // This isn't really correct, but otherwise we can end up in a case where
+    // we constrain inference variables by selecting a single predicate, when
+    // we need to stay general. See issue #91762.
+    let (_, predicate_own_substs) = obligation.predicate.trait_ref_and_own_substs(selcx.tcx());
+    if predicate_own_substs.iter().any(|g| g.has_infer_types_or_consts()) {
+        return Ok(ProjectedTy::NoProgress(
```
Comment (nikomatsakis) [inline compiler/rustc_trait_selection/src/traits/project.rs:1092]: I think we want to treat this the same way we would treat an ambiguous set of candidates (i.e., `Err(ProjectionTyError::TooManyCandidates)`)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by jackh726** [comment]
Updated

@rustbot ready

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**